lib/pbench/server/indexing_tarballs.py

```diff
@@ -103,69 +117,71 @@ class Index:
+
+        # Index context name
+        self.name: str = name
 
-    def collect_tb(self):
-        """Collect tarballs that needs indexing"""
+        # An Elasticsearch status Report object
+        self.report: Optional[Report] = None
 
-        # find -L $ARCHIVE/*/$linksrc -name '*.tar.xz' -printf "%s\t%p\n" 2>/dev/null | sort -n > $list
-        tarballs = []
+        # We'll use a FileTree in order to find the INCOMING directory for the
+        # unpacked tarballs.
+        self.filetree: FileTree = FileTree(idxctx.config, idxctx.logger)
+
+        # Manage synchronization between components
+        self.sync: Sync = Sync(idxctx.logger, "index")  # Build a sync object
+
+    def collect_tb(self) -> Tuple[int, List[TarballData]]:
+        """Collect tarballs that need indexing
+
+        This looks for Datasets marked to require indexing, and returns a tuple
+        consisting of an error number from the 'error_code' attribute and a
+        list of tarball descriptions.
+
+        Returns:
+            A list of tarball descriptions sorted by tarball size.
+        """
+
+        tarballs: List[TarballData] = []
         idxctx = self.idxctx
         error_code = self.error_code
         try:
-            tb_glob = os.path.join(self.archive, "*", self.linksrc, "*.tar.xz")
-            for tb in glob.iglob(tb_glob):
-                try:
-                    rp = Path(tb).resolve(strict=True)
-                except OSError:
-                    idxctx.logger.warning("{} does not resolve to a real path", tb)
-                    quarantine(self.qdir, idxctx.logger, tb)
-                    continue
-                controller_path = rp.parent
-                controller = controller_path.name
-                archive_path = controller_path.parent
-                if str(archive_path) != str(self.archive):
-                    idxctx.logger.warning(
-                        "For tar ball {}, original home is not {}", tb, self.archive
-                    )
-                    quarantine(self.qdir, idxctx.logger, tb)
-                    continue
-                if not Path(f"{rp}.md5").is_file():
-                    idxctx.logger.warning("Missing .md5 file for {}", tb)
-                    quarantine(self.qdir, idxctx.logger, tb)
-                    # Audit should pick up missing .md5 file in ARCHIVE directory.
+            datasets = self.sync.next(self.operation)
+            for dataset in datasets:
+                tb = Metadata.getvalue(dataset, Metadata.TARBALL_PATH)
+                if not tb:
+                    self.sync.error(dataset, f"{dataset} does not have a tarball-path")
                     continue
+
+                # We use the tarball size to sort the list so that smaller
+                # datasets don't get stalled behind large ones. (Though the
+                # size is at best an indirect indication of "indexing cost",
+                # it's relatively easy to compute.)
```

Comment:
While this is an interesting factoid, I don't think it provides any value here greater than the existing comment at line 178, and it duplicates verbiage elsewhere  (e.g., at line 73).  The problem is that there is no code here which is doing anything like sorting or scheduling, and, if we were to change our mind about how we present the `tarballs` list, it's unlikely we'll remember to change this comment.

So, I recommend that you remove this text; if you really want to keep it, you should put it at line 200, where we actually do the sorting.